File: Algorithms/Python/searching/linear_search_recursive.py
from __future__ import annotations


def linear_search_recursive(arr: list[int], target: int, index: int = 0) -> int:
    if index == len(arr):
        return -1
    if arr[index] == target:
        return index
    return linear_search_recursive(arr, target, index + 1)


if __name__ == "__main__":
    numbers = [1, 2, 3, 9, 4, 5, 6]
    print(linear_search_recursive(numbers, 9))
    print(linear_search_recursive(numbers, 7))

File: Algorithms/C/searching/linear_search_recursive.c
#include <stdio.h>

int linear_search_recursive(const int arr[], int target, int length)
{
    if (length == 0)
    {
        return -1;
    }
    if (arr[length - 1] == target)
    {
        return length - 1;
    }
    return linear_search_recursive(arr, target, length - 1);
}

int main()
{
    int arr[] = {1, 2, 3, 4, 5};
    int target = 3;
    int length = sizeof(arr) / sizeof(arr[0]);
    int result = linear_search_recursive(arr, target, length);

    printf("Index at: %d", result);

    return 0;
}
